### ISSUE TITLE
dev branch: Kallisto only Workflow that can process paired files

### ISSUE THREAD

**Issue by JohnHadish** [SystemsGenetics/GEMmaker#252]
This issue pertains to the dev branch witch currently contains the new dsl2 branch.

Only Kallisto can process paired end reads.
Demonstration:

1) Clone copy of GEMmaker to get the demo directory. Switch to dev branch
```
git clone git@github.com:SystemsGenetics/GEMmaker.git
cd GEMmaker
git checkout dev
```

2) Demonstrate that Hisat works fine with non-paired SRAs. We will use the first arabidopsis example from [How to Launch GEMmaker](https://gemmaker.readthedocs.io/en/dev/execution.html#how-to-launch-gemmaker)
```
# Put non-paired SRA in SRA.txt
echo SRR1058270 > SRAs.txt 

# Run Hisat2
nextflow run main.nf -profile singularity \
  --pipeline hisat2 \
  --hisat2_base_name CORG \
  --hisat2_index_dir assets/demo/references/CORG.genome.Hisat2.indexed/ \
  --hisat2_gtf_file assets/demo/references/CORG.transcripts.gtf \
  --sras SRAs.txt

# *This Completes*
```

2) Now we try the same command with a paired end file. We will use the file from the [CORG example](https://github.com/SystemsGenetics/GEMmaker/blob/master/assets/demo/SRA_IDs.txt)
```
# Put paired end file in SRAs.txt
echo SRR649944 > SRAs.txt 
# Run Hisat2
nextflow run main.nf -profile singularity \
  --pipeline hisat2 \
  --hisat2_base_name CORG \
  --hisat2_index_dir assets/demo/references/CORG.genome.Hisat2.indexed/ \
  --hisat2_gtf_file assets/demo/references/CORG.transcripts.gtf \
  --sras SRAs.txt

# *This fails*
```

3) We can run the same file with Kallisto, and it will succeed:
```
# Remove Hisat2 results and work dirs
rm work results -r

nextflow run main.nf -profile singularity \
  --pipeline kallisto \
  --kallisto_index_path assets/demo/references/CORG.transcripts.Kallisto.indexed \
  --sras SRAs.txt

# *This succeeds*
```

This can be repeated with other files, both local and remote.

===
Here is the error message for Hisat2:
```
Error executing process > 'GEMmaker:trimmomatic (SRX218012)'

Caused by:
  Process `GEMmaker:trimmomatic (SRX218012)` terminated with an error exit status (1)

Command executed:

  echo "#TRACE sample_id=SRX218012"
  echo "#TRACE n_cpus=4"
  echo "#TRACE minlen=0.7"
  echo "#TRACE leading=3"
  echo "#TRACE trailing=6"
  echo "#TRACE slidingwindow=4:15"
  echo "#TRACE fasta_lines=`cat fasta_adapter.txt | wc -l`"
  echo "#TRACE fastq_lines=`cat *.fastq | wc -l`"
  
  # convert the incoming FASTQ file list to an array
  read -a fastq_files <<< SRX218012_1.fastq SRX218012_2.fastq
  
  # This script calculates average length of fastq files.
  total=0
  # This if statement checks if the data is single or paired data, and checks length accordingly
  # This script returns 1 number, which can be used for the minlen in trimmomatic
  if [ ${#fastq_files[@]} == 2 ]; then
    for fastq in "${fastq_files[@]}"; do
      cat="cat $fastq"
      if [[ $fastq =~ .gz$ ]]; then
        cat="zcat $fastq"
      fi
      a=`$cat | awk 'NR%4 == 2 {lengths[length($0)]++} END {for (l in lengths) {print l, lengths[l]}}'         | sort         | awk '{ print $0, $1*$2}'         | awk -v var="0.7" '{ SUM += $3 } { SUM2 += $2 } END { printf("%.0f", SUM / SUM2 * var)} '`
    total=($a + $total)
    done
    total=( $total / 2 )
    minlen=$total
  else
    cat="cat ${fastq_files[0]}"
    if [[ ${fastq_files[0]} =~ .gz$ ]]; then
      cat="zcat ${fastq_files[0]}"
    fi
    minlen=`$cat | awk 'NR%4 == 2 {lengths[length($0)]++} END {for (l in lengths) {print l, lengths[l]}}'          | sort         | awk '{ print $0, $1*$2}'         | awk -v var="0.7" '{ SUM += $3 } { SUM2 += $2 } END { printf("%.0f", SUM / SUM2 * var)} '`
  fi
  if [ ${#fastq_files[@]} == 2 ]; then
    trimmomatic         PE         -threads 4         ${fastq_files[0]}         ${fastq_files[1]}         SRX218012_1p_trim.fastq         SRX218012_1u_trim.fastq         SRX218012_2p_trim.fastq         SRX218012_2u_trim.fastq         ILLUMINACLIP:fasta_adapter.txt:2:40:15         LEADING:3         TRAILING:6         SLIDINGWINDOW:4:15         MINLEN:"$minlen" > SRX218012.trim.log 2>&1
  else
    trimmomatic         SE         -threads 4         ${fastq_files[0]}         SRX218012_1u_trim.fastq         ILLUMINACLIP:fasta_adapter.txt:2:40:15         LEADING:3         TRAILING:6         SLIDINGWINDOW:4:15         MINLEN:"$minlen" > SRX218012.trim.log 2>&1
  fi

Command exit status:
  1

Command output:
  #TRACE sample_id=SRX218012
  #TRACE n_cpus=4
  #TRACE minlen=0.7
  #TRACE leading=3
  #TRACE trailing=6
  #TRACE slidingwindow=4:15
  #TRACE fasta_lines=130
  #TRACE fastq_lines=5480

Command error:
  .command.sh: line 12: read: `SRX218012_2.fastq': not a valid identifier

Work dir:
  /home/jah/Documents/test/GEMmaker/work/fe/b8cdd7186eb2af71f374da3b069a90

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line


WARN: To render the execution DAG in the required format it is required to install Graphviz -- See http://www.graphviz.org for more info.

```

**Comment by JohnHadish** [comment]
Figured it out, the bash function "read" is only avalible in the kallisto docker image:
```
singularity run https://depot.galaxyproject.org/singularity/salmon:1.5.2--h84f40af_0 read 
/usr/local/env-execute: line 3: exec: read: not found
singularity run https://depot.galaxyproject.org/singularity/kallisto:0.46.2--h4f7b962_1 read
WARNING: Skipping mount /usr/local/var/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container
FATAL:   "read": executable file not found in $PATH
```

**Comment by JohnHadish** [comment]
The issue is with this bash command and how it is interpreted within the docker images:
```
read -a fastq_files <<< ${fastq_files}
```

I believe that the interpretation of the `<<<` is not consistent

**Comment by spficklin** [comment]
Fixed in PR #253
